File: C++/Circular Queue.cpp
#include<iostream>
using namespace std;

const int MAX_SIZE = 5;

typedef int element;
typedef struct{
    element array[MAX_SIZE];
    int front, rear;
} Queue;

void init(Queue *q)
{
    q->front = q->rear = 0;
}

bool empty(Queue *q)
{
    if(q->front == q->rear)
        return true;
    return false;
}

void print(Queue *q)
{
    if (empty(q))
    {
        cout <<"Queue is empty \n";
        return;
    }
    
    cout << "(front: "<< q->front << " rear: "<< q->rear <<")\n";
    cout << "q: [ ";
    
    // Initial idx is front
    int idx = q->front;
    do
    {
        idx = (idx+1) % MAX_SIZE;
        cout << q->array[idx]<< " ";
        if (idx == q->rear)
            break;
    } while(idx!= q->front);
    cout<<"]\n";
}

bool full(Queue *q)
{
    if ((q->rear + 1) % MAX_SIZE == q->front)
        return true;
    return false;
}

void enqueue(Queue *q, element item)
{
    if (full(q))
    {
        cout <<"Queue is saturated \n";
        return;
    }
    // maintaining the shape of a circular queue using the operator %
    q->rear = (q->rear + 1) % MAX_SIZE;
    q->array[q->rear] = item;
}

void deque(Queue *q)
{
    if(empty(q))
    {
        cout << "Queue is empty";
        return;
    }
    
    q->front = (q->front + 1) % MAX_SIZE;
    int result = q->array[q->front];
    cout<< result<< " was removed\n";
    
  
}

int main()
{
    Queue q;
    
    init(&q);
    cout << "===Input data===\n\n";
    int num = 10;
    while(!full(&q))
    {
        enqueue(&q, num);
        print(&q);
        num += 20;
        cout << "\n";
    }
    
    cout << "\n===Delete data===\n\n";
    while(!empty(&q))
    {
        deque(&q);
        print(&q);
        cout << "\n";
    }
    
    return 0;
}
/*
 ===Input data===

 (front: 0 rear: 1)
 q: [ 10 ]

 (front: 0 rear: 2)
 q: [ 10 30 ]

 (front: 0 rear: 3)
 q: [ 10 30 50 ]

 (front: 0 rear: 4)
 q: [ 10 30 50 70 ]


 ===Delete data===

 10 was removed
 (front: 1 rear: 4)
 q: [ 30 50 70 ]

 30 was removed
 (front: 2 rear: 4)
 q: [ 50 70 ]

 50 was removed
 (front: 3 rear: 4)
 q: [ 70 ]

 70 was removed
 Queue is empty
 */

File: Python/Circular Queue.py
MAX_SIZE = 10

class CircularQueue:
    def __init__(self):
        self.front = 0
        self.rear = 0
        self.array = [None] * MAX_SIZE

    def empty(self):
        if self.front == self.rear:
            return True
        return False

    def full(self):
        if self.front == (self.rear + 1) % MAX_SIZE:
            return True
        return False

    def print(self):
        if self.empty():
            print("Circular Queue is empty")
            return
        print("(front: {} rear: {})".format(self.front, self.rear))
        ans = []
        if self.front < self.rear:
            ans = self.array[self.front+1:self.rear]
        else:
            ans = self.array[self.front+1:MAX_SIZE] + self.array[0:self.rear + 1]
        print("q: ", ans)

    def enqueue(self, item):
        if self.full():
            print("Circular Queue is saturated")
            return
        self.rear = (self.rear + 1) % MAX_SIZE
        self.array[self.rear] = item

    def deque(self):
        if self.empty():
            print("Circular Queue is empty")
            return
        self.front = (self.front + 1) % MAX_SIZE
        result = self.array[self.front]
        print(result, " was removed")


if __name__ == '__main__':
    q = CircularQueue()
    num = 1

    while not q.full():  # (front: 0 rear: 9)
        q.enqueue(num)
        num += 2

    q.print()            # q:  [1, 3, 5, 7, 9, 11, 13, 15]

    for _ in range(3):   # remove 1, 3 ,5 -> (front: 3 rear: 9)
        q.deque()

    q.print()            # q:  [7, 9, 11, 13, 15]
